pulsar/internal/crypto/producer_encryptor.go

```diff
@@ -0,0 +1,78 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package crypto
+
+import (
+	"fmt"
+
+	"github.com/apache/pulsar-client-go/pulsar/crypto"
+	"github.com/apache/pulsar-client-go/pulsar/log"
+)
+
+type producerEncryptor struct {
+	keys                        []string
+	keyReader                   crypto.KeyReader
+	messageCrypto               crypto.MessageCrypto
+	logger                      log.Logger
+	producerCryptoFailureAction int
+}
+
+func NewProducerEncryptor(keys []string,
+	keyReader crypto.KeyReader,
+	messageCrypto crypto.MessageCrypto,
+	producerCryptoFailureAction int,
+	logger log.Logger) Encryptor {
+	return &producerEncryptor{
+		keys:                        keys,
+		keyReader:                   keyReader,
+		messageCrypto:               messageCrypto,
+		logger:                      logger,
+		producerCryptoFailureAction: producerCryptoFailureAction,
+	}
+}
+
+// Encrypt producer encryptor
+func (e *producerEncryptor) Encrypt(payload []byte, msgMetadata crypto.MessageMetadataSupplier) ([]byte, error) {
+	// encryption is enabled but KeyReader interface is not implemented
+	if e.keyReader == nil {
```

Comment:
Should this be detected and an error raised while setting up the producer?

Reply:
done